docs/proposals/rescheduling.md

```diff
@@ -0,0 +1,496 @@
+  requests/decisions (e.g. just as an eviction, an eviction with a hint about where to
+  reschedule, or as an eviction paired with a specific binding), how does the system
+  implement these requests, does the rescheduler take into account the second-order
+  effects of decisions (e.g. whether an evicted pod will reschedule, will cause
+  a preemption when it reschedules, etc.), does the rescheduler execute multi-step plans
+  (e.g. evict two pods at the same time with the intent of moving one into the space
+  vacated by the other, or even more complex plans)
+
+Additional musings on the rescheduling design space can be found [here](rescheduler.md).
+
+## Design proposal
+
+The key mechanisms and components of the proposed design are priority, preemption,
+disruption budgets, the `/evict` subresource, and the rescheduler.
+
+### Priority
+
+#### Motivation
+
+[Quality of Service levels](resource-qos.md) allow Kubernetes to safely overcommit nodes.
+Just as it is useful to overcommit nodes to increase node-level utilization, it is useful
+to overcommit clusters to increase cluster-level utilization.  Scheduling priority (which
+we abbreviate as *priority*, in combination with disruption budgets (described in the
+next section), allows Kubernetes to safely overcommit clusters much as QoS levels allow
+it to safely overcommit nodes.
+
+Today, cluster sharing among users, workload types, etc. is regulated via the
+[quota](../admin/resourcequota/README.md) mechanism. When allocating quota, a cluster
+administrator has two choices: (1) the sum of the quotas is less than or equal to the
+capacity of the cluster, or (2) the sum of the quotas is greater than the capacity of the
+cluster (that is, the cluster is overcommitted).  (1) is likely to lead to cluster
+under-utilization, while (2) is unsafe in the sense that someone's pods may go pending
+indefinitely even though they are still within their quota. Priority makes cluster
+overcommitment (i.e. case (2)) safe by allowing users and/or administrators to identify
+which pods should be allowed to run, and which should go pending, when demand for cluster
+resources exceeds supply to due to cluster overcommitment.
+
+Priority is also useful in some special-cause scenarios, such as ensuring that system
+DaemonSets can always schedule and reschedule onto every node where they want to run
+(assuming they are given the highest priority), e.g. see #21767.
+
+#### Specifying priorities
+
+We propose to add a required `Priority` field to `PodSpec`. Its value type is string, and
+the cluster administrator defines a total ordering on these strings (for example
+`Critical`, `Normal`, `Preemptible`). We choose string instead of integer so that it is
+easy for an administrator to add new priority levels in between existing levels, to
+encourage thinking about priority in terms of user intent and avoid magic numbers, and to
+make the internal implementation more flexible.
```

Comment:
> This is related to the question @mqliang asked later about whether preemption should be allowed at all across namespaces (my cop-out was that this should be a scheduler configuration pattern, but I thought most people would want the answer to be "yes").

If we allow preemption across namespaces, there is a problem about ResourceQuota:
>
Generally, there are two viewpoints about this question, both of them sounds reasonable.
>
Viewpoint 1: can not preempt across namespaces
Reason : A namespace indicating a user community in k8s. All users are equal, so namespace has no priority, just pod has priority(i.e. Pod of DaemonSet has the highest priority, Pod of Job has the lowest priority). If a user want run more high-priority Pods when he can not preempt his low-priority pod in further step, he should apply for more ResourceQuota(for example, pay more money), instead of preempt low-priority pods of other users.
>
Someone might imagine that it does not decrease availability even if low-priority pods could be preempted by other user's high-priority pods, since we have "DisruptionBudget", which could ensure availability, but I feel nervous about it, for example: As a user, I get a ResourceQuota(I may paid some money to get one) which allow me create 5 Pods, so service provider(k8s) must ensure I could always successfully create and run 5 Pods, right(I paid the money)? If unfortunately, my pods could be preempted by others' high-priority Pods, I feel it's hard to accept.
>
Viewpoint 2: can preempt across namespaces
Reason : In such a case, as a user, I get a ResourceQuota which allow me create AT MOST 10 Pods, but just ensure 5 pods's availability(DisruptionBudget is 50%). So I could create at most 10 pods, service provider(k8s) will make sure there are 5 pods are always Running. As a bonus, 5 pods I created could be Running if cluster resource is sufficient. But service provider(k8s) doesn't give any promise about the latter 5 pods: they may be preempted by other users' high-priority pods at any time.
>
So, if we want preempt across namespaces, ResourceQuota should take DisruptionBudget into account too. ResourceQuota should not only restrict the number of all kinds of Pods in one namespace, but also restrict the number of "effective" Pod. "effective" Pods means the Pods protected by DisruptionBudget, those Pods will never become unavailable.

All my assumption was based on my understanding of ResourceQuota: as a user, I pay some money to get a ResourceQuota which allow me create 5 Pods, then service provider(k8s) should always ensure I could create and run 5 pods. If my understanding was wrong, feel free to correct me.

This also related to the "namespace(group)/local(user) priority" problem. My understanding is: namespace(user) may have priority, but such a priority means how many resource they can get, a high-priority namespace(user) could get a ResourceQuota which allow him create more Pods(or more high-priority Pods).